### ISSUE TITLE
Credentials Persistence Location

### ISSUE THREAD

**Issue by CodeMan99** [aws/aws-toolkit-vscode#3138]
## Problem

While using [devcontainers](https://code.visualstudio.com/docs/devcontainers/containers), the default location of `$HOME/.aws/credentials` is usually temporary. Containers are normally automatically deleted after a configuration change.

## Suggestion One

Allow IAM credentials to be read from the environment: `$AWS_ACCESS_KEY_ID` & `$AWS_SECRET_ACCESS_KEY`.

This enables users manual control over persistence of their credentials.

## Suggestion Two

Save IAM credentials to the VS code settings for the extension. Likely under the keys: `aws.iam.accessKeyId` & `aws.iam.secretAccessKey`.

This allows users to have some environments with the extension and other environments without.

**Comment by JadenSimon** [comment]
Hey @CodeMan99 

Suggestion one should already be available in the latest extension version, however, we might be missing "auto-connect" for environment variables. You should see `env:variables` in the quick pick when running the "Connect to AWS" command.

Suggestion two is interesting. I don't think we want to allow credentials in the user settings, however, it might make sense to allow users to specify the location of the credentials/config files. Would something like that work for you?

**Comment by CodeMan99** [comment]
> Suggestion one should already be available in the latest extension version, however, we might be missing "auto-connect" for environment variables. You should see `env:variables` in the quick pick when running the "Connect to AWS" command.

Do mean here?

![image](https://user-images.githubusercontent.com/6373201/215913147-4c275177-ecfe-4287-b3b2-a723d321d9b6.png)



> Suggestion two is interesting. I don't think we want to allow credentials in the user settings, however, it might make sense to allow users to specify the location of the credentials/config files. Would something like that work for you?

Do not really understand the apprehension over the user settings. The `$HOME/.aws/credentials` path is no more secure.

--

I suppose I could create a docker volume that mounted at `$HOME/.aws`.

**Comment by JadenSimon** [comment]
> Do mean here?

Yes if you set `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` prior to launching VS Code there should be an `env:variables` option. If there isn't then that could be a bug or some other issue that we'd need to look into.

> Do not really understand the apprehension over the user settings. The $HOME/.aws/credentials path is no more secure.

You're right, a file on disk isn't really any more secure than the settings. It's less about how secure the location is but instead more about how isolated the data is from other things. More isolation means less chance of accidentally leaking things, at least in my opinion. Maybe the isolation isn't _that_ important, but I tend to be overly cautious when it comes to auth.

**Comment by CodeMan99** [comment]
> Yes if you set `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` prior to launching VS Code there should be an `env:variables` option. If there isn't then that could be a bug or some other issue that we'd need to look into.

I do have these values set from `terminal.integrated.env.linux`. Which I suppose is the problem. They are isolated to the integrated terminal and don't appear on a normal login shell.

I'm not sure where to even set the values so that an extension can read the values...

> Maybe the isolation isn't _that_ important, but I tend to be overly cautious when it comes to auth.

Ah, you know, that's somewhat fair. Isolation is usually good.

**Comment by CodeMan99** [comment]
Tomorrow I will give the [remoteEnv](https://code.visualstudio.com/remote/advancedcontainers/environment-variables) configuration object a go.

**Comment by JadenSimon** [comment]
> I do have these values set from terminal.integrated.env.linux. Which I suppose is the problem. They are isolated to the integrated terminal and don't appear on a normal login shell.
>
> I'm not sure where to even set the values so that an extension can read the values...

Yup, `terminal.integrated.env.linux`  only affects the integrated terminal. Using local environment variables is a bit annoying because you have to completely close out of VS Code and re-launch with the variables set like so:
```
AWS_ACCESS_KEY_ID=... AWS_SECRET_ACCESS_KEY=... code
```

> Tomorrow I will give the [remoteEnv](https://code.visualstudio.com/remote/advancedcontainers/environment-variables) configuration object a go.

That should work! Of course, it won't be very user-friendly if the credentials change frequently. But that's another topic.

**Comment by justinmk3** [comment]
> VS code settings for the extension. Likely under the keys: `aws.iam.accessKeyId` & `aws.iam.secretAccessKey`.
> 
> This allows users to have some environments with the extension and other environments without.

vscode settings "scope" rules can be tricky. And if the remote machine (devcontainer) changes, non-global (i.e. remote-specific) settings won't carry over to a new one, which (I believe) defeats the use-case here?

Improving AWS Toolkit understanding of environment variables is tracked in:

- https://github.com/aws/aws-toolkit-vscode/issues/2029
- related but different: https://github.com/aws/aws-toolkit-vscode/issues/1351
-
